File: 2020/old/discomplexity/x/src/x/graphics/vertices.cc
#include "vertices.hh"
#include "application.hh"

namespace x
{
    namespace graphics
    {
        vertices::vertices()
        :   __mode(x::graphics::vertices::none),
            __type(x::graphics::vertices::none),
            __adjacency(x::graphics::vertices::none),
            __id(0),
            __vertices()
        {
        }

        vertices::vertices(const vertices & o)
        :   __mode(o.__mode),
            __type(o.__type),
            __adjacency(o.__adjacency),
            __id(o.__id),
            __vertices(o.__vertices)
        {
        }

        vertices::vertices(vertices && o) noexcept
        :   __mode(o.__mode),
            __type(o.__type),
            __adjacency(o.__adjacency),
            __id(o.__id),
            __vertices(x::move(o.__vertices))
        {
            o.__mode = x::graphics::vertices::none;
            o.__type = x::graphics::vertices::none;
            o.__adjacency = x::graphics::vertices::none;
            o.__id = 0;
        }

        vertices::~vertices()
        {
        }

        vertices & vertices::operator=(const vertices & o)
        {
            if(&o != this)
            {
                __mode = o.__mode;
                __type = o.__type;
                __adjacency = o.__adjacency;
                __id = o.__id;
                __vertices = o.__vertices;
            }
            return *this;
        }

        vertices & vertices::operator=(vertices && o) noexcept
        {
            if(&o != this)
            {
                __mode = o.__mode;
                __type = o.__type;
                __adjacency = o.__adjacency;
                __id = o.__id;
                __vertices = x::move(o.__vertices);

                o.__mode = x::graphics::vertices::none;
                o.__type = x::graphics::vertices::none;
                o.__adjacency = x::graphics::vertices::none;
                o.__id = 0;
            }
            return *this;
        }

        void vertices::serialize(x::graphics::buffer & buffer) const
        {
            if(buffer.size() == 0)
            {
                for(x::uint64 i = 0; i < __vertices.size(); i++)
                {
                    const x::graphics::vertex & vertex = __vertices[i];
                    buffer.push(vertex.position());
                    buffer.push(vertex.normal());
                    buffer.push(vertex.color());
                    buffer.push(vertex.texture());
                    buffer.push(vertex.tangent());
                    buffer.push(vertex.binormal());
                }
                // x::console::out << "buffer size => " << buffer.size() << x::endl;
            }
        }
    }
}

File: 2020/old/discomplexity/x/js/x/graphics/vertices.js

export default class Vertices
{
    static get none(){ return 0; }
    static get point(){ return 1; }
    static get line(){ return 2; }
    static get patch(){ return 3; }
    static get triangle(){ return 4; }

    static get strip(){ return 1; }
    static get loop(){ return 2; }
    static get fan(){ return 3; }

    static get adjacent(){ return 1; }

    static modeFromStr(s)
    {
        if(s === "triangle")
        {
            return Vertices.triangle;
        }
        else
        {
            throw new Error();
        }
        return s;
    }

    constructor()
    {
        this.__mode = Vertices.none;
        this.__type = Vertices.none;
        this.__adjacency = Vertices.none;
        this.__id = Vertices.none;
        this.__vertices = [];
    }

    push(v)
    {
        this.__vertices.push(v);
    }

    set(v)
    {
        this.__vertices = v;
    }

    serialize(buffer)
    {
        let array = [];
        let total = 0;

        this.__vertices.forEach(vertex => {
            let result = [];
            array.push(new Float32Array(result.concat(vertex.position,
                                                      vertex.normal,
                                                      vertex.color,
                                                      vertex.texture,
                                                      vertex.tangent,
                                                      vertex.binormal)));
            total += array[array.length - 1].length;
        });
        let o = new Float32Array(total);
        total = 0;
        array.forEach(value => {
            o.set(value, total);
            total += value.length;
        });
        buffer.data = o;
    }
    get size(){ return this.__vertices.length; }
    set mode(v){ this.__mode = v; }
}
